addons/point_of_sale/static/src/app/store/pos_store.js

```diff
@@ -236,13 +236,46 @@ export class PosStore extends Reactive {
         this.config.iface_printers = !!this.unwatched.printers.length;
 
         // Monitor product pricelist
+        this.models["product.product"].addEventListener(
+            "create",
+            this.computeProductPricelistCache.bind(this)
+        );
+        this.models["product.pricelist.item"].addEventListener(
+            "create",
+            this.computeProductPricelistCache.bind(this)
+        );
+
         this.computeProductPricelistCache();
     }
 
-    computeProductPricelistCache() {
+    computeProductPricelistCache(data) {
+        // This function is called via the addEventListener callback initiated in the
+        // processServerData function when new products or pricelists are loaded into the PoS.
+        // It caches the heavy pricelist calculation when there are many products and pricelists.
         const date = DateTime.now();
-        const products = this.models["product.product"].getAll();
-        const pricelistItems = this.models["product.pricelist.item"].getAll();
+        let pricelistItems = this.models["product.pricelist.item"].getAll();
+        let products = this.models["product.product"].getAll();
+
+        if (data && data.length > 0) {
+            if (data[0].model.modelName === "product.product") {
+                products = data;
+            }
+
+            if (data[0].model.modelName === "product.pricelist.item") {
+                pricelistItems = data;
+
+                const productTplIds = pricelistItems.map((pi) => pi.raw.product_tmpl_id);
+                const productIds = pricelistItems.map((p) => p.raw.product_id);
+
+                if (productTplIds) {
+                    products = products.filter((p) =>
+                        productTplIds.includes(p.raw.product_tmpl_id)
+                    );
+                } else if (productIds) {
+                    products = products.filter((p) => productIds.includes(p.raw.product_id));
+                }
```

Comment:
```suggestion
```
remove this, we should loop to all the products because for pricelist.item that don't have both 'product_id' and 'product_tmpl_id', it will be applicable to any product.